### PR TITLE
feat(#major): Ribbon Finance ; add Ribbon Finance subgraph

### PR DESCRIPTION
**Pull request by shashwatS22**

### CODE REVIEW

**Review by dmelotik — COMMENTED** [review]
Nice work @shashwatS22 ! I made some comments, and some apply to multiple places if it is a common pattern. Let me know when you made changes or responded to my comments if you disagree. Then I will finish the review.

Few notes:
- Can you add a README.md and use another protocol to follow the same format. And make any notes (a bullet list) for anything you had to do out of the ordinary or weird events you noticed.
- Do you have a deployment of this subgraph you can share? A link to it here would be great: https://okgraph.xyz/
- Not sure if you have seen this, but Ribbon has a subgraph you can reference: https://github.com/ribbon-finance/ribbon-subgraph/tree/main

File: subgraphs/ribbon-finance/src/common/initalizers.ts [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+import { Address, ethereum, BigInt } from "@graphprotocol/graph-ts";
+import { ERC20 as ERC20Contract } from "../../generated/ETHCallV2/ERC20";
+import { LiquidityGaugeV5 as GaugeContract } from "../../generated/rETHThetaGauge/LiquidityGaugeV5";
+import { RibbonThetaVaultWithSwap as VaultContract } from "../../generated/ETHCallV2/RibbonThetaVaultWithSwap";
+
+export function getOrCreateToken(
+  address: Address,
+  block: ethereum.Block,
+  vault: Address = constants.NULL.TYPE_ADDRESS,
+  isOToken: bool = false
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/initalizers.ts:29]: OToken was initially the name I used for option tokens which were used as pools when we were using the generic schema, but now we moved to YA schema and oToken is used as output tokens,  I'll change the name to isOutputToken.

File: subgraphs/ribbon-finance/src/modules/Deposit.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  );
+
+  const depositAmountUSD = depositAmount
+    .toBigDecimal()
+    .times(inputTokenPrice.usdPrice)
+    .div(inputTokenDecimals.toBigDecimal());
+
+  vault.outputTokenSupply = utils.getOutputTokenSupply(vaultAddress, block);
+
+  const totalValue = utils.readValue<BigInt>(
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/modules/Deposit.ts:104]: Thanks!

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      "address": "0x16772a7f4a3ca291C21B8AcE76F9332dDFfbb5Ef",
+      "startBlock": 12369554
+    },
+    {
+      "name": "RibbonYearnETHPutV1",
+      "address": "0x8FE74471F198E426e96bE65f40EeD1F8BA96e54f",
+      "startBlock": 12696707
+    }
+  ],
+  "liquidityGauges": [
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json:89]: I will shift this to using the liquidity gauge controller.

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json [hunk 4]
```diff
@@ -0,0 +1,152 @@
+{
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json:1]: I didn't know that we were only using V2 and Earn vaults, I'll make the necessary changes.

File: subgraphs/ribbon-finance/src/common/utils.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    vaultContract.try_totalBalance(),
+    constants.BIGINT_ZERO
+  ).divDecimal(constants.BIGINT_TEN.pow(vaultDecimals as u8).toBigDecimal());
+
+  const totalOptions = readValue(
+    vaultContract.try_totalSupply(),
+    constants.BIGINT_ZERO
+  ).divDecimal(constants.BIGINT_TEN.pow(vaultDecimals as u8).toBigDecimal());
+
+  const pricePerShare = totalTokensDeposits.div(totalOptions);
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/utils.ts:99]: I think we can remove this code now, as we are removing the V1 vaults because this logic was explicitly for the V1 vaults as they don't have a method to get the pricePerShare directly.

File: subgraphs/ribbon-finance/src/mappings/vault.ts [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+export function handleRollToNextOption(call: RollToNextOption): void {
+  const vaultAddress = call.to;
+  const block = call.block;
+  rollToNextOption(vaultAddress, block);
+  log.warning("[HandleRollToNextOption] transaction hash {}", [
+    call.transaction.hash.toHexString(),
+  ]);
+}
+
+export function handleOpenShort(event: OpenShort): void {
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/mappings/vault.ts:216]: I will remove these handlers.

File: subgraphs/ribbon-finance/src/mappings/vault.ts [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+    "[DistributePremium] transaction hash {} premium {} premiumUSD {}",
+    [
+      event.transaction.hash.toHexString(),
+      premium.toString(),
+      premiumUSD.toString(),
+    ]
+  );
+}
+
+export function handleRollToNextOption(call: RollToNextOption): void {
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/mappings/vault.ts:207]: Earlier the option tokens were used as the output tokens, so this handler was used to assign the new output token weekly whenever the vault moves to a new cycle, but now the logic is changed so we can remove this handler.

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+          #Treasury
+        - event: DistributePremium(uint256,uint256[],address[],uint256)
+          handler: handleDistributePremium  
+        - event: CollectManagementFee(uint256,uint256,indexed address)
+          handler: handleCollectManagementFee
+        - event: CollectPerformanceFee(uint256,uint256,indexed address)
+          handler: handleCollectPerformanceFee  
+                
+
+      callHandlers:
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml:116]: I'll keep a note of it while developing future subgraphs.

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json:1]: @shashwats22 we went back and forth internally. And I apologize but we do want the V1 vaults in there. We will keep the treasury out. And it will be good to make note of this in the readme (the scope of the vaults). Thanks for the patience

[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/deployments/ribbon-finance-eth/configurations.json:1]: No worries, I'll add back the v1 vaults.

File: subgraphs/ribbon-finance/src/modules/Deposit.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+  if (totalValue.notEqual(constants.BIGINT_ZERO)) {
+    vault.totalValueLockedUSD = vault.inputTokenBalance
+      .toBigDecimal()
+      .times(inputTokenPrice.usdPrice)
+      .div(inputTokenDecimals.toBigDecimal());
+  }
+
+  const pricePerShare = utils.getVaultPricePerShare(vaultAddress);
+  vault.pricePerShare = pricePerShare.times(inputTokenDecimals.toBigDecimal());
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/modules/Deposit.ts]: I was getting the wrong value for price per share, so I was just testing an approach. Fixed this in the current push.

File: subgraphs/ribbon-finance/src/common/constants.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+export const BIGINT_ZERO = BigInt.fromI32(0);
+export const BIGINT_ONE = BigInt.fromI32(1);
+export const BIGINT_TEN = BigInt.fromI32(10);
+export const BIGINT_HUNDRED = BigInt.fromI32(100);
+
+export const BIGDECIMAL_ZERO = new BigDecimal(BIGINT_ZERO);
+export const BIGDECIMAL_HUNDRED = BigDecimal.fromString("100");
+export const BIGDECIMAL_NEGATIVE_ONE = BigDecimal.fromString("-1");
+
+export const PANCAKE_FACTORY_ADDRESS = Address.fromString(
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/constants.ts]: These were some extra addresses that I missed, I have removed them in the current commit.

File: subgraphs/ribbon-finance/src/modules/Withdraw.ts [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  const vaultDecimals = constants.BIGINT_TEN.pow(vault._decimals as u8);
+
+  const withdrawAmountUSD = withdrawAmount
+    .toBigDecimal()
+    .div(vaultDecimals.toBigDecimal())
+    .times(inputTokenPrice.usdPrice);
+
+  vault.outputTokenSupply = utils.getOutputTokenSupply(vaultAddress, block);
+
+  const totalValue = utils.readValue<BigInt>(
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/modules/Withdraw.ts]: I have made the change and merged the Deposit.ts file and Withdraw.ts file into the Transaction.ts file.

File: subgraphs/ribbon-finance/schema.graphql [hunk 12]
```diff
@@ -0,0 +1,737 @@
...
+  " The number of decimal places this token uses, default to 18 "
+  decimals: Int!
+
+  " Optional field to track the price of a token, mostly for caching purposes "
+  lastPriceUSD: BigDecimal
+
+  " Optional field to track the block number of the last token price "
+  lastPriceBlockNumber: BigInt
+
+  _isOtoken: Boolean
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/schema.graphql]: I have changed it to _isOutputToken

File: subgraphs/ribbon-finance/src/common/utils.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+export function getVaultPricePerShare(vaultAddress: Address): BigDecimal {
+  const vaultContract = VaultContract.bind(vaultAddress);
+  const vaultDecimals = readValue(vaultContract.try_decimals(), 18);
+
+  const vaultPricePerShare = readValue(
+    vaultContract.try_pricePerShare(),
+    constants.BIGINT_ZERO
+  ).divDecimal(constants.BIGINT_TEN.pow(vaultDecimals as u8).toBigDecimal());
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/utils.ts]: I have added this utility function to the code.

**Comment by dmelotik** [comment]
@shashwatS22 can you add your deployment links to the header comment?

**Comment by shashwatS22** [comment]
> Nice work @shashwatS22 ! I made some comments, and some apply to multiple places if it is a common pattern. Let me know when you made changes or responded to my comments if you disagree. Then I will finish the review.
> 
> Few notes:
> 
> * Can you add a README.md and use another protocol to follow the same format. And make any notes (a bullet list) for anything you had to do out of the ordinary or weird events you noticed.
> * Do you have a deployment of this subgraph you can share? A link to it here would be great: https://okgraph.xyz/
> * Not sure if you have seen this, but Ribbon has a subgraph you can reference: https://github.com/ribbon-finance/ribbon-subgraph/tree/main

> I have modified the old README.md that was there in the file.
> Here is the okgraph.xyz link: https://okgraph.xyz/?q=QmNQRg2YGM8bsfhaajYuN1xWZGJ594de27j3kmE2njoyeE
> I used the Ribbon official subgraph as a reference for some of the events.

**Comment by shashwatS22** [comment]
> @shashwatS22 can you add your deployment links to the header comment?

@dmelotik I have added the deployment links, I redeployed the code so I was waiting for it to sync completely before adding the link, its synced now.

**Review by dmelotik — COMMENTED** [review]
Few more comments, but I went through everything this time. So when they are fixed I think it should be ready to go!

**Comment by dmelotik** [comment]
Also, if it is not too much work we want to add the avalanche theta vaults.

This is actually a really good example of why we want to put contract addresses in the config files and not hardcode the template manifest 🧠

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+          handler: handlePayOptionYield
+        - event: PurchaseOption(uint256,indexed address)
+          handler: handlePurchaseOption
+          #V1
+        - event: Withdraw(indexed address,uint256,uint256,uint256)
+          handler: handleWithdrawWithFee
+
+  {{/vaults}}
+    
+  {{#liquidityGauges}}
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml:98]: We need this because the V1 vaults don't have a gauge controller.

File: subgraphs/ribbon-finance/src/mappings/vault.ts [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+    vaultAddress,
+    false
+  );
+  const totalFeeUSD = utils
+    .bigIntToBigDecimal(totalFee, vault._decimals)
+    .times(vaultAsset.lastPriceUSD!);
+
+  updateRevenueSnapshots(vault, constants.BIGDECIMAL_ZERO, totalFeeUSD, block);
+  updateVaultSnapshots(vaultAddress, event.block);
+  log.warning("[CollectVaultFees] transaction hash {}", [
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/mappings/vault.ts:175]: Removed them.

File: subgraphs/ribbon-finance/package.json [hunk 16]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "ribbon-generic",
+  "license": "UNLICENSED",
+  "scripts": {
+    "codegen": "graph codegen",
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/package.json:5]: I have not used the messari cli yet, I'll try this.

File: subgraphs/ribbon-finance/src/common/initializers.ts [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+    vault = new VaultStore(vaultAddress.toHexString());
+
+    const vaultContract = VaultContract.bind(vaultAddress);
+    vault.name = utils.readValue<string>(vaultContract.try_name(), "");
+    vault.symbol = utils.readValue<string>(vaultContract.try_symbol(), "");
+    vault.protocol = constants.PROTOCOL_ID;
+
+    vault._decimals = utils.readValue(vaultContract.try_decimals(), 0);
+
+    vault.depositLimit = utils.readValue<BigInt>(
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/initializers.ts:267]: Added new events!

File: subgraphs/ribbon-finance/src/modules/Rewards.ts [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    false
+  );
+
+  if (!vault.rewardTokens) {
+    vault.rewardTokens = [];
+  }
+
+  const rewardTokens = vault.rewardTokens!;
+  if (!rewardTokens.includes(rewardToken.id)) {
+    rewardTokens.push(rewardToken.id);
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/modules/Rewards.ts:180]: I'll keep a note of this while working on future subgraphs.

File: subgraphs/ribbon-finance/src/common/initializers.ts [hunk 19]
```diff
@@ -0,0 +1,507 @@
...
+      token._isOutputToken = true;
+      token._vaultId = vault.toHexString();
+
+      token.lastPriceUSD = utils.getOutputTokenPriceUSD(vault, block);
+      token.lastPriceBlockNumber = block.number;
+    }
+    token.save();
+  }
+
+  if (
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/initializers.ts:70]: I removed the condition for updating the price once per hour, now I am updating the price whenever getOrCreateToken is called and changed the prices call in updateVaultTVL to getOrCreateToken so that it also updates the price.

File: subgraphs/ribbon-finance/src/common/initializers.ts [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+
+    vault.cumulativeSupplySideRevenueUSD = constants.BIGDECIMAL_ZERO;
+    vault.cumulativeProtocolSideRevenueUSD = constants.BIGDECIMAL_ZERO;
+    vault.cumulativeTotalRevenueUSD = constants.BIGDECIMAL_ZERO;
+
+    const withdrawlFeeId =
+      utils.enumToPrefix(constants.VaultFeeType.WITHDRAWAL_FEE) +
+      vaultAddress.toHexString();
+
+    getOrCreateFee(withdrawlFeeId, constants.VaultFeeType.WITHDRAWAL_FEE);
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/initializers.ts:321]: I added the code to update the fees, I think I missed updating the fees.

File: subgraphs/ribbon-finance/src/common/constants.ts [hunk 21]
```diff
@@ -0,0 +1,116 @@
+import { Address, BigDecimal, BigInt } from "@graphprotocol/graph-ts";
+
+////////////////////
+///// Versions /////
+////////////////////
+
+export const PROTOCOL_NAME = "Ribbon Finance";
+export const PROTOCOL_SLUG = "ribbon-finance";
+export const PROTOCOL_SCHEMA_VERSION = "1.3.0";
```
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/src/common/constants.ts]: I have used the versions file and updated the code.

File: package.json [hunk 22]
```diff
@@ -23,7 +23,7 @@
   },
   "scripts": {
     "prettier": "prettier --loglevel silent --write \"**/*.{md,jsx,tsx,ts,json,html,css,js}\"",
-    "prettier-check": "prettier --check",
```
Comment (dmelotik) [inline package.json:26]: undo this change

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml [hunk 23]
```diff
@@ -0,0 +1,547 @@
...
+          file: ./abis/RibbonThetaVaultWithSwap.json
+        - name: LiquidityGaugeV5
+          file: ./abis/LiquidityGaugeV5.json
+        - name: RibbonVaultPauser
+          file: ./abis/RibbonVaultPauser.json
+        - name: ERC20
+          file: ./abis/ERC20.json
+        - name: GaugeController
+          file: ./abis/GaugeController.json
+        ###########################################
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml:431]: The prices are not used in the handler, you can remove those ABIs from here

File: subgraphs/ribbon-finance/src/common/initializers.ts [hunk 24]
```diff
@@ -0,0 +1,489 @@
...
+    token.name = utils.readValue<string>(contract.try_name(), "");
+    token.symbol = utils.readValue<string>(contract.try_symbol(), "");
+    token.decimals = utils.readValue<i32>(
+      contract.try_decimals(),
+      constants.DEFAULT_DECIMALS.toI32() as u8
+    );
+    token._vaultId = vault.toHexString();
+    token.lastPriceBlockNumber = block.number;
+    if (isOutputToken) {
+      token._isOutputToken = true;
```
Comment (dmelotik) [inline subgraphs/ribbon-finance/src/common/initializers.ts:53]: This is also redundant, you can set `token._isOutputToken = isOutputToken` outside of the if block.

btw, I am just being picky here, but we should aim to make our code as succinct as possible for easier readability and maintenance in the future

File: subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml (hunk 23 above)
[the remark this answers is not in the document]
Comment (shashwatS22) [reply subgraphs/ribbon-finance/protocols/ribbon-finance/config/templates/ribbon.finance.template.yaml]: Removed them.

**Comment by shashwatS22** [comment]
> 2 more thing.
> 
> * We had to add ribbon-finance to the deployment.json earlier, so it is actually in there 2x rn. [here](https://github.com/messari/subgraphs/blob/master/deployment/deployment.json#L3355)
> * One thing you can do is rebase onto the latest master branch and then it will be there. I would honestly suggest to just undo your edits on deployment.json though. (easier)
> * Not sure if you saw this, but we did the QA on Ribbon. There are a few pricing things on ethereum to fix ([Ribbon Finance - MAINNET QA #1727](https://github.com/messari/subgraphs/issues/1727))
> 
> Let me know if you want to do the QA fixes in a separate PR to keep it cleaner. The subgraph will be `prod` ready after that PR.

I have undone the deployment.json changes and fixed the QA issues. Thanks!

**Review by dmelotik — APPROVED** [review]
Nice work! Merging
